File: samples/python/coordination/peer_to_peer_delegation.py
from langchain_anthropic import ChatAnthropic

llm = ChatAnthropic(model="claude-haiku-4-5-20251001")

# Agent registry — simulates A2A service discovery
AGENT_REGISTRY: dict[str, dict] = {
    "agent_pdf_analyzer": {
        "id": "agent_pdf_analyzer",
        "capabilities": ["pdf_analysis", "document_extraction"],
        "endpoint": "local://agent_b",
    },
    "agent_web_scraper": {
        "id": "agent_web_scraper",
        "capabilities": ["web_scraping", "html_parsing"],
        "endpoint": "local://agent_c",
    },
}

def discover_agent(required_capability: str) -> dict | None:
    for agent_id, info in AGENT_REGISTRY.items():
        if required_capability in info["capabilities"]:
            print(f"[Registry] Found agent '{agent_id}' with capability '{required_capability}'")
            return info
    return None

# Agent B — the peer that handles PDF analysis
def agent_b_handle(task: str) -> str:
    print(f"[Agent B] Received delegation for: {task}")
    result = llm.invoke(
        f"You are a PDF analysis agent. Analyze this task and provide a 2-sentence result:\n{task}"
    )
    return result.content.strip()

# Simulate A2A call based on discovered endpoint
def delegate_to_peer(peer_info: dict, task: str) -> str:
    endpoint = peer_info["endpoint"]
    print(f"[A2A] Delegating to endpoint: {endpoint}")
    if endpoint == "local://agent_b":
        return agent_b_handle(task)
    return "Unknown peer endpoint."

# Agent A — discovers and delegates
def agent_a(task: str) -> str:
    print(f"[Agent A] Received task: {task}")
    required = "pdf_analysis"
    print(f"[Agent A] I lack capability '{required}', searching registry...")
    peer = discover_agent(required)
    if peer is None:
        return "[Agent A] No capable peer found. Cannot complete task."
    result = delegate_to_peer(peer, task)
    print(f"[Agent A] Got result from peer: {result}")
    return result

def main():
    task = "Extract the financial summary and key metrics from the Q1 2025 earnings PDF report."
    print("=== Peer-to-Peer Delegation (A2A) ===\n")
    final_result = agent_a(task)
    print(f"\n[Final Result]\n{final_result}")

if __name__ == "__main__":
    main()

File: samples/go/coordination/peer_to_peer_delegation.go
package coordination

import (
	"fmt"
	"strings"
)

// P2PAgent handles tasks or delegates to the next agent in the chain.
type P2PAgent struct {
	name string
	next *P2PAgent
	// canHandle returns true if this agent can handle the task.
	canHandle func(task string) bool
}

func (a *P2PAgent) Handle(task string) string {
	if a.canHandle(task) {
		return fmt.Sprintf("[%s] handled: %s", a.name, task)
	}
	if a.next != nil {
		fmt.Printf("  [%s] cannot handle, delegating to %s\n", a.name, a.next.name)
		return a.next.Handle(task)
	}
	return fmt.Sprintf("[%s] no handler found for: %s", a.name, task)
}

// RunPeerToPeerDelegation demonstrates the Peer-to-Peer Delegation pattern.
func RunPeerToPeerDelegation() {
	fmt.Println("=== Peer-to-Peer Delegation Pattern ===")

	agentC := &P2PAgent{
		name: "AgentC",
		canHandle: func(task string) bool {
			return strings.Contains(task, "translate")
		},
	}
	agentB := &P2PAgent{
		name: "AgentB",
		next: agentC,
		canHandle: func(task string) bool {
			return strings.Contains(task, "language")
		},
	}
	agentA := &P2PAgent{
		name: "AgentA",
		next: agentB,
		canHandle: func(task string) bool {
			return strings.Contains(task, "2+2")
		},
	}

	tasks := []string{
		"calculate 2+2",
		"fix language grammar",
		"translate this sentence to French",
	}

	for _, task := range tasks {
		fmt.Printf("Task: %q\n", task)
		result := agentA.Handle(task)
		fmt.Printf("  Result: %s\n", result)
	}
	fmt.Println()
}
